src/github.com/matrix-org/dendrite/clientapi/sync/requestpool.go

```diff
@@ -54,34 +75,70 @@ func (rp *RequestPool) OnIncomingSyncRequest(req *http.Request) util.JSONRespons
 		"timeout": timeout,
 	}).Info("Incoming /sync request")
 
-	res, err := rp.currentSyncForUser(syncReq)
-	if err != nil {
-		return httputil.LogThenError(req, err)
-	}
-	return util.JSONResponse{
-		Code: 200,
-		JSON: res,
+	// Fork off 2 goroutines: one to do the work, and one to serve as a timeout.
+	// Whichever returns first is the one we will serve back to the client.
+	// TODO: Currently this means that cpu work is timed, which may not be what we want long term.
+	timeoutChan := make(chan struct{})
+	timer := time.AfterFunc(timeout, func() {
+		close(timeoutChan) // signal that the timeout has expired
+	})
+
+	done := make(chan util.JSONResponse)
+	go func() {
+		syncData, err := rp.currentSyncForUser(syncReq)
```

Comment:
do we not need to cancel this after a timeout? Otherwise we're going to end up stacking them up forever if a user keeps calling sync but never gets any events?